0005-snarkyjs-ecdsa-sha.md

```diff
@@ -0,0 +1,179 @@
+
+However, the existing `Poseidon` API could be deprecated in favour of the new `Hash` namespace.
+
+```ts
+const Hash = {
+  Poseidon: {
+    hash: (xs: UInt8[]) => "digest",
+  },
+
+  // ..
+
+  SHA3_224: {
+    hash: (xs: UInt8[]) => "digest",
+  },
+
+  Keccak: {
+    hash: (xs: UInt8[]) => "digest",
+  },
+};
+```
+
+Developers would then be able to simply import the `Hash` namespace into their project and use all available hash functions.
+
+```ts
+import { Hash } from "snarkyjs";
+
+Hash.Poseidon.hash(xs);
+Hash.Keccak.hash(xs);
+// ..
+
+// deprecated in favor of Hash.Poseidon.hash
+Poseidon.hash(xs);
+```
+
+It is important to mention that we have a lot of freedom in the the implementation details of the API, as the gates and gadgets are already implemented in OCaml and just need to exposed and wrapped by a developer friendly API.
+
+One goal of the API is to design an easy to use interface for developers as well as maintaining consistency with already existing primitives (Poseidon in this case, Schnorr in the case of ECDSA).
+The usage of both Keccak/SHA3 and Poseidon is the same: The user passes an array of Field elements into the hash function and the output is returned. One important detail is that Keccak/SHA3 only accepts an array of Field element that each are no larger than 1 byte. We will also provide additional helper functions that allow conversion between Field element arrays and hexadecimal encoded strings.
+
+```ts
+function fieldBytesFromHex(hex: string): UInt8[];
+function hexToFieldBytes(xs: UInt8[]): string;
+```
+
+Overall, exposing new gadgets and gates follow a strict pattern that has been used all over in the SnarkyJS bindings layer. As an example, the [Poseidon](https://github.com/o1-labs/snarkyjs-bindings/blob/main/ocaml/lib/snarky_js_bindings_lib.ml#L386) implementation behaves similarly. From the point of view of SnarkyJS, these gadgets are just another set of function calls to OCaml.
+
+## ECDSA
+
+**TODO** this will be a seperate PR stacked on top of this one
+
+**Evergreen, wide-sweeping Details**
+
+**Ephemeral details**
+
+## Test plan and functional requirements
+
+## SHA3/Keccak
+
+In order to test the implementation of SHA3 and Keccak in SnarkyJS, we will follow the testing approach we already apply to other gadgets and gates.
+This includes testing the out-of- and in-snark variants using our testing framework, as well as adding a SHA3 and Keccak regression test. The regression tests will also include a set of predetermined digests to make sure that the algorithm doesn't unexpectedly change over time (similar to the tests implemented for the OCaml gadget).
+
+In addition to that, we should provide a dedicated integration test that handles SHA3/Keccak hashing within a smart contract (proving enabled). This will allow us to not only provide developers with an example, but also ensure that SHA3 and Keccak proofs can be generated.
```

Comment:
- Are we concerned about performance testing? 
  - For example, hashing large amounts of data or rapidly making hash requests.
- Are we concerned about collision testing?
- Are we concerned about security testing? 
  - Possible/known vulnerabilities, cryptographic weakness, etc.
- Are we concerned about usability testing?
  - In a form of gathering the feedback from a small group of developers to improve the design and usability of the API before releasing it.

---

- The RFC indicates that hashing with SHA3 and Keccak is more expensive than using Poseidon. Performance metrics, such as time and memory usage, should be gathered and analyzed to provide developers with a clear understanding of the subject.

Reply:
Good question - @querolita to what extend has the original crypto implementation be tested? Do you have any recommendations to what we should test, if at all?

Reply:
The tests performed to the implementation were:
- Checking different output lengths and different configurations of the padding (NIST's for SHA3 and submission's for Keccak)
- Checking different length outputs with different numbers of blocks
- Edge cases of padding to make sure that a whole new block is created, and messages that already look padded
- Checking different endianness of inputs and outputs
- Check that non-byte length inputs cannot be passed
- Can reuse constraint system for the same circuit structure
- Cannot reuse constraint system for different output lengths
- Cannot reuse constraint system for same output length but different padding
- Cannot reuse constraint system for different endianness

I would create similar tests for SnarkyJS, but probably more high level since you don't have access to low level details such as constraint system creation (that I am aware of). In particular, I would make sure that inputs which are not "correct formatted hex values" (meaning an odd number of hex digits) is not passed to the function. And of course making sure that all combinations actually produce the expected value (pairs of output lengths and padding types), for some random values. I wouldn't go much deeper than that regarding the snarkyjs level alone.

In that sense, we are not testing against collision, just comparing against official outputs of the hash function on those given inputs, nor performance testing (all we know is 1 block of encryption, which is enough for 1080bits of input data, fits in 15k kimchi rows, thus 4 full blocks can be hashed without chunking), we haven't looked for keccak's potential vulnerabilities, and we have not gathered usability info from users other than ourselves.

Reply:
@shimkiv does this sounds good to you? Is there anything else you would change here?

Reply:
> we will follow the testing approach we already apply to other gadgets and gates

It will be good to add some implementation example links here.

> will also include a set of predetermined digests

Can we please be more specific here? What kind of set? Is it standardised set? If not then how we determine the set to be good enough?

> similar to the tests implemented for the OCaml gadget

Can we please include links here?

> We will include a range of edge cases in the tests (e.g. empty input, zero, etc).

It will be good if we can be more explicit here and at least highlight the major cases we're aware of and that will be included into the testing plan.

> In that sense, we are not testing against collision, just comparing against official outputs of the hash function on those given inputs, nor performance testing (all we know is 1 block of encryption, which is enough for 1080bits of input data, fits in 15k kimchi rows, thus 4 full blocks can be hashed without chunking), we haven't looked for keccak's potential vulnerabilities, and we have not gathered usability info from users other than ourselves.

If we agree on this then it would be good if we can add listed items into the "out of the scope" testing section with some rationale.